File: pack2/sender.py
import socket

from pack2.address import Address


class Sender:

    def __init__(self, address: Address):
        self._address = address.ip
        self._port = address.port
        self._connected = False
        self._socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM, socket.IPPROTO_TCP)

    def connect(self):
        if not self._connected:
            try:
                self._socket.connect((self._address, self._port))
                self._connected = True
            except socket.error:
                raise ConnectionRefusedError('Failed to connect to server')

    def close(self):
        if self._connected:
            self._connected = False
            self._socket.close()

    def send(self, data: bytes):
        if not self._connected:
            self.connect()
        self._socket.send(data)

    def response(self) -> bytes:
        return self._socket.recv(1)

File: pack2/golangVers/socket_connection/sender.go
package socket_connection

import (
	"net"
	"strconv"
)

type Sender interface {
	Close() error
	Send([]byte) error
	Response() []byte
}

type SocketSender struct {
	address string
	port    int
	socket  net.Conn
}

func NewSocketSender(address string, port int) *SocketSender {
	res := &SocketSender{address: address, port: port}
	var err error
	res.socket, err = net.Dial("tcp", address+":"+strconv.Itoa(res.port))
	if err != nil {
		panic(err)
	}
	return res
}

func (s *SocketSender) Close() error {
	return s.socket.Close()
}

func (s *SocketSender) Send(data []byte) error {
	sent, err := s.socket.Write(data)
	for sent != len(data) {
		if err != nil {
			return err
		}
		var nowSent = 0
		nowSent, err = s.socket.Write(data[sent:])
		sent += nowSent
	}
	return err
}

func (s *SocketSender) Response() []byte {
	data := make([]byte, 10)
	_, err := s.socket.Read(data)
	if err != nil {
		panic(err)
	}
	return data
}
